installer: external assets

bastimeyer opened the pull request:
Resolves #2872
and upgrades FFMpeg to 4.2.2 (Zeranoe)

### Changes

1. Rewrite makeinstaller.sh
   - Move removed plugins list into a text file and read from it
   - Add log and err functions
   - Define build dependencies
   - Set repo root directory instead of assuming it
   - Clean up build directory before building the installer
   - Set dist dir to same STREAMLINK_DIST_DIR
   - Suppress warning/error messages from inkscape and imagemagick
   - Replace some strings in pynsist config with variables
   - Update python license text according to the currently used version
2. Download assets from streamlink/streamlink-assets
   - Download metadata from latest assets release
   - Download assets into cache dir (useful for re-building locally)
   - Compare checksums
   - Extract specific files following the metadata
3. Finally remove binary files from the repo

beardypig commented on script/makeinstaller.sh:
```diff
@@ -1,62 +1,81 @@
-mkdir -p "${build_dir}" "${dist_dir}" "${nsis_dir}" "${files_dir}" "${icons_dir}"
 
-echo "Building streamlink-${STREAMLINK_VERSION} package..." 1>&2
+removed_plugins_file="${ROOT}/src/streamlink/plugins/_removed.txt"
```
I personally prefer `.removed.txt` as the name for that file, or even just `.removed`

bastimeyer replied:
The idea was to have the file sorted at the top. A leading dot would hide it in most *nix file managers by default.

beardypig replied:
Hidden is the best place for it, imo.

bastimeyer replied:
Sure, let's change it. Will update after some more feedback and issue resolving.

beardypig commented on script/makeinstaller.sh:
```diff
@@ -1,62 +1,81 @@
-do
-    touch "${build_dir_plugins}/$i.py"
-done
+while read -r pluginname; do
+    touch "${build_dir_plugins}/${pluginname}.py"
+done < <(sed -e 's/[[:space:]]*[#;].*//; /^[[:space:]]*$/d' "${removed_plugins_file}")
```
I think it's a bit more readable to use `readarray`, it's a Bash 4 command but that shouldn't be a problem. 

```suggestion
readarray -t removed_plugins < "${removed_plugins_file}"
for pluginname in "${removed_plugins[@]}"; do
    touch "${build_dir_plugins}/${pluginname}.py"
done
```

bastimeyer replied:
Does readarray remove comments with a leading `#` or `;`?

beardypig replied:
True, you got comments in there uglying it up :p 
I made a similar change on another branch, but didn't put comments in the plugin list - I wanted to make a script to check to make sure a removed plugin was added to that file.

bastimeyer replied:
I don't have anything against removing the comments. I just included them for anyone randomly reading the file.

> I wanted to make a script to check to make sure a removed plugin was added to that file

That script/test could also remove the comments while readin from the file. :smiley: 

But ok, let me change this, too.

beardypig replied:
Nah, leave the comments they are useful. `grep -x` fixes everything.
```
while read -r rplugin; do echo grep -Fxq "${rplugin}" src/streamlink/plugins/_removed.txt; done < <(git diff --cached --name-only --diff-filter=D master | grep -oP '^src/streamlink/plugins/(\K.*)(?=.py)');
```

bastimeyer replied:
Ok, restored again. For some reason Github doesn't update the review comments and I have to refresh the page.

beardypig commented on script/makeinstaller.sh:
```diff
@@ -1,62 +1,81 @@
 
-set -e # stop on error
+ROOT=$(git rev-parse --show-toplevel 2>/dev/null || realpath "$(dirname "$(readlink -f "${0}")")/..")
+cd "${ROOT}"
```
```suggestion
pushd "${ROOT}"
```

And `popd` on exit :)

bastimeyer replied:
The entire script is executed in a subshell, so it shouldn't matter whether `cd` or `pushd/popd` is used, right?

beardypig replied:
True, force of habit

beardypig commented on script/makeinstaller.sh:
```diff
@@ -262,17 +280,51 @@ StrCmp \$0 \${ffmpeg} "" +2
+    IFS=" " read -r filename sourcedir targetdir \
+        < <(jq -r ".[$i] | \"\(.filename) \(.sourcedir) \(.targetdir)\"" <<< "${assets_data}")
+    unzip -qq "${cache_dir}/${filename}" -d "${TEMP}"
+    while IFS=" " read -r from to; do
+        log "Copying ${sourcedir}/${from} to ${targetdir}/${to}"
+        install -D -T "${TEMP}/${sourcedir}/${from}" "${files_dir}/${targetdir}/${to}"
+    done < <(jq -r ".[$i].files[] | \"\(.from) \(.to)\"" <<< "${assets_data}")
```
I find it a bit nicer to pipe into while, it appears to work the same for me. 

```suggestion
    jq -r ".[$i].files[] | \"\(.from) \(.to)\"" <<< "${assets_data}" | while read -r from to; do
        log "Copying ${sourcedir}/${from} to ${targetdir}/${to}"
        install -D -T "${TEMP}/${sourcedir}/${from}" "${files_dir}/${targetdir}/${to}"
    done
```

bastimeyer replied:
Piping into while makes it run in a sub-shell, which is what I usually try to avoid.
https://stackoverflow.com/questions/13726764/while-loop-subshell-dilemma-in-bash

Not relevant in this particular case, but I tried to keep the same style throughout the assets and makeinstaller scripts. In the assets script, I ran into an issue because of piping into while.

beardypig replied:
I like it because it isolates the variables, keeps them in their scope. It's easier to read too.

beardypig replied:
Leave it how it is then, it's mostly a style issue for me anyway - I prefer to have the input to at the start of the loop rather than at the end :)
Bash script are hard to read as it is, probably this script would be better off in Python ;D

bastimeyer replied:
I'll leave it. Otherwise the line gets too long and moving the while statement with the pipe symbol into the next line with added indentation makes it look even more weird.

beardypig commented on script/makeinstaller.sh:
```diff
@@ -262,17 +280,51 @@ StrCmp \$0 \${ffmpeg} "" +2
+log "Extracting assets"
+TEMP=$(mktemp -d) && trap "rm -rf ${TEMP}" EXIT || exit 255
+for ((i=$(jq length <<< "${assets_data}") - 1; i >= 0; --i)); do
+    IFS=" " read -r filename sourcedir targetdir \
```
Do you need to set `$IFS` here?

bastimeyer replied:
I've explicitly set it on all read commands in case the user has set it. I don't think it's really needed, but better safe than sorry. We can export it at the start of the script though, if you prefer it that way.

beardypig replied:
Doesn't bash set it as `" \t\n"` at the start of the script automatically?

bastimeyer replied:
Yep, not necessary and can be removed as it doesn't inherit the value. The bash man page is only talking about default values, so wasn't 100% sure.

bastimeyer commented:
The log output on Github actions is so useless. It's scrambling the lines for some reason, even in the raw log.

- https://github.com/streamlink/streamlink/pull/2916/checks?check_run_id=595990158#step:7:10
- https://pipelines.actions.githubusercontent.com/eFmcCyrUFDy9z9353ozAyQf6s8IEKyozIRvlSLnsHPO1JgSIpe/_apis/pipelines/1/runs/115/signedlogcontent/16?urlExpires=2020-04-17T16%3A40%3A41.9663179Z&urlSigningMethod=HMACV1&urlSignature=OMJc2qdc9T5spL2ug1vb4%2FUkgedJNRT2ga8Ij3%2BVBE8%3D

bastimeyer commented:
What about the python license file? Do we need to include it? I feel like this is pynsist's job.

beardypig commented:
It is pynsist's job. But it probably should be included, depends on what the licence says :)

bastimeyer commented:
https://github.com/streamlink/streamlink/pull/2916/files#diff-6871e707e71cec9606ed3e5534cc4516

Ok, that should probably not be deleted, but instead fixed up.

bastimeyer commented:
Restored `win32/LICENSE.txt`, fixed typo, fixed python version (would probably be wise turning it into a template in the future when the python version gets bumped again) and removed the redundant RTMPDump license text.

Still not sure how to feel about the removal of the license files of the libs bundled by FFmpeg. The zeranoe build archive should include these, but it doesn't. Maybe include this?
https://ffmpeg.zeranoe.com/builds/readme/win64/static/ffmpeg-4.2.2-win64-static-readme.txt
Would require a quick update of the assets repo and a fix here, as it always expects zip archives.

back-to commented on win32/LICENSE.txt:
```diff
@@ -1,662 +1,320 @@
+8. By copying, installing or otherwise using Python 3.6.6, Licensee agrees
+   to be bound by the terms and conditions of this License Agreement.
+
+pbs License
```
>  pbs License

can be removed

Ref https://github.com/streamlink/streamlink/pull/441

bastimeyer replied:
True, will remove that.

Does anyone know why some dependencies have their licenses included here and some not?

bastimeyer commented:
I've submitted streamlink/streamlink-assets#2 earlier today which should resolve the ffmpeg license issue. This needs to be merged first before I can push my changes here.

Btw, also did a bit of research regarding the Python license and from what it looks like, it's simply missing from the `python-3.6.6-embed-win32.zip` file `pynsist` downloads + extracts. The `3.8.2` zip archive includes a license file, so once the installer's python version gets bumped (not in this PR), the custom license text can be removed.

bastimeyer commented:
The ffmpeg build info is now included and the assets related log output a bit more verbose:
https://github.com/streamlink/streamlink/runs/599289704?check_suite_focus=true#step:7:308

That should be it regarding the external assets and their licenses.

Btw, `pynsist` 2.5 - we're currently using 2.4 - has some optimizations regarding the assembly of the executable launchers. Those are now done at build time and the result is that the install directory is much cleaner. Also includes high-DPI scaling fixes. Could be explored in another PR.

beardypig commented:
Have you made all the changes you want to make for this PR @bastimeyer?

bastimeyer commented:
Unless you've spotted another issue, I think it's ready to be merged.

beardypig commented:
Looks good to me, let's see if @back-to can spot anything :D

bastimeyer commented:
Ehm, do we even include Streamlink's license in the installer? I don't think so...

> Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

back-to commented on script/makeinstaller.sh:
```diff
@@ -1,72 +1,91 @@
 version=${STREAMLINK_VERSION}
 entry_point=streamlink_cli.main:main
 icon=${icons_dir}/icon.ico
+license_file=${ROOT}/LICENSE
```
would be better as a `.txt` file on windows

gravyboat commented:
Changes look good to me. 👍
